_posts/2019-05-21-i18n-messages.adoc

```diff
@@ -0,0 +1,327 @@
+
+Messages can be received programmatically via the `Messages` Interface. It provides an API to fetch a translated String through the localization mechanism of CUBA for the locale of the current user.
+
+TIP: In the example code instead of using the `Messages` Interface, the `MessageBundle` Interface is used. It is a new shorthand mechanism introduced in CUBA 7.0, that fetches a particular message from the message pack of the calling Java class. This way it is not required to provide the package for every translation API request.
+
+
+The first localized message for pre-filling the description attribute is created as follows:
+
+
+.VisitBrowse.java
+[source,java]
+----
+public class VisitBrowse extends StandardLookup<Visit> {
+
+    @Inject
+    private Notifications notifications;
+
+    @Inject
+    private MessageBundle messageBundle;
+
+    // ...
+
+    private void createVisitForPet(Pet pet) {
+        screenBuilders.editor(visitsTable)
+                .newEntity()
+                .withInitializer(visit -> {
+                    visit.setDescription(
+                        regularCheckupDescriptionContent(pet) // <1>
+                    );
+                    visit.setPet(pet);
+                })
+                .withScreenClass(RegularCheckup.class)
+                .withLaunchMode(OpenMode.DIALOG)
+                .withAfterCloseListener(event -> showRegularCheckupCreatedNotification(pet))
+                .build()
+                .show();
+    }
+
+    private String regularCheckupDescriptionContent(Pet pet) {
+        return messageBundle.formatMessage( // <2>
+                "regularCheckupContent", // <3>
+                pet.getName(), // <4>
+                pet.getIdentificationNumber()
+        );
+    }
+----
+<1> the description of the visit instance will be set on entity creation
+<2> the `messageBundle` bean is responsible for fetching the translated message
+<3> `regularCheckupContent` is the key of the message to be translated
+<4> the parameters `pet.getName()` and `pet.getIdentificationNumber()` are passed to the translation as dynamic parts
+
+For the dynamic parts of a message, the translation should contain the placeholder `%s` like in this definition:
+
+.messages_de.properties
+[source,properties]
+----
+regularCheckupContent=Kontrolluntersuchung für %s (%s)\n\n
+
+Bestandteile: \n
+- Krallen-Kontrolle\n
+- Temperatur messen\n
+- Zahn-Check
+----
+
+This code results into the following description value of the regular check:
+
+image::i18n-messages/editor-with-programmatic-created-i10n-messages.png[align="center"]
+
+
+The second message should be displayed as a notification once the Regular Checkup was successfully created. There is
+ the method `showRegularCheckupCreatedNotification` in the Visit browser, that will be called once the instance is created:
+
+
+.VisitBrowse.java
+[source,java]
+----
+public class VisitBrowse extends StandardLookup<Visit> {
+
```

Comment:
No need for many empty lines, it looks not great on the resulting page.

Reply:
👍 - btw do you have an idea why the `// ...` is always placed without indentation? I have tried it several different ways, but it never seems to be working :-/